src/main/java/net/neoforged/neoforge/fluids/FluidUnitSpec.java

```diff
@@ -0,0 +1,183 @@
+        this.map = data;
+        this.listForTooltip = map.entrySet().stream()
+                .filter(e -> e.getValue().showInTooltip)
+                .sorted(Comparator.<Map.Entry<FluidUnit, UnitInfo>, Integer>comparing(e -> e.getValue().value).reversed())
+                .map(Map.Entry::getKey)
+                .toList();
+        this.defaultUnit = defaultUnit;
+        this.smallestUnitDecimals = smallestUnitDecimals;
+    }
+
+    /**
+     * Verify that if bucket or mb are present, they must have the correct values
+     */
+    private static DataResult<Map<FluidUnit, UnitInfo>> checkFixedUnits(Map<FluidUnit, UnitInfo> map) {
+        for (Map.Entry<FluidUnit, UnitInfo> entry : map.entrySet()) {
+            if (entry.getKey() == FluidUnits.BUCKET && entry.getValue().value != FluidType.BUCKET_VOLUME)
+                return DataResult.error(() -> "bucket must be defined as " + FluidType.BUCKET_VOLUME);
+            if (entry.getKey() == FluidUnits.MB && entry.getValue().value != 1)
+                return DataResult.error(() -> "mb must be defined as 1");
+        }
+        return DataResult.success(map);
+    }
+
+    /**
+     * Gets a list of all units in this spec..
+     */
+    public Set<FluidUnit> units() {
+        return map.keySet();
+    }
+
+    /**
+     * Gets a default unit that may be used by mods that want to display the quantity
+     * in a limited amount of space as a single number.
+     */
+    private FluidUnit defaultUnit() {
+        return this.defaultUnit;
+    }
+
+    /**
+     * Gets the quantity in millibuckets associated with a given unit.
+     */
+    public @Nullable Integer getValue(FluidUnit unit) {
+        UnitInfo info = map.get(unit);
+        if (info != null) return info.value;
+        if (unit == FluidUnits.BUCKET)
+            return FluidType.BUCKET_VOLUME;
+        if (unit == FluidUnits.MB)
+            return 1;
+        return null;
+    }
+
+    /**
+     * Gets a component suitable for use in a tooltip representing the quantity,
+     * e.g. "1 block, 3 ingots"
+     */
+    public List<Component> getTooltipComponents(long amount) {
+        if (amount == 0) return List.of(Component.translatable("fluid_unit.empty"));
+        List<Component> results = new ArrayList<>();
+        for (int i = 0; i < listForTooltip.size(); i++) {
+            FluidUnit unit = listForTooltip.get(i);
+            @Nullable
+            Integer unitValue = getValue(unit);
+            if (unitValue == null) continue; // shouldn't happen
+            long amountUnit = amount / unitValue;
+            amount = amount % unitValue;
+            if (i < listForTooltip.size() - 1 || amount == 0) {
+                if (amountUnit != 0)
+                    results.add(Component.translatable("fluid_unit.number_and_unit_format", // "%s %s"
+                            amountUnit, unit.getDescriptionId(amountUnit != 1)));
```

Comment:
There are plenty of languages for which this way of pluralising is wrong.

`com.ibm.icu.text.PluralFormat` is available.

Reply:
If/when #1269 gets merged, you can simplify this.

```
  "fluid_unit.c.block": "%n(one{# Block} other{# Blocks})",
```

and then simply `results.add(Component.translatable(unit.getDescriptionId(), amountUnit))`.

BTW, consider using getTranslationKey(), as that's what's used in (at least some) other places.

Reply:
`getDescriptionId` is the MojMap name for that method in vanilla classes. IIRC `getTranslationKey` was the MCP name before we moved to mojmaps, and not every area got switched over for vanilla consistency.